### PR TITLE
Feat Assignment 18 File Stream

### PR DESCRIPTION
**Pull request by aristovinceselvanathan**
In this assignment, I have understood the different stream that available for the file operations in c#. Task I have read the 1gb of the file using the file stream and buffered stream to compare they are performance. Asynchronous concept also applied on the file reading and writing operations without waiting for the operation to be completed. Parallel Library used to read and process the files simultaneously.

### CODE REVIEW

File: src/Assignment18/FileDataProcessor/DataProcessor.cs [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+                stopwatch.Restart();
+                ColorfulMessage("\nPress Enter to read the file using the buffered stream", ConsoleColor.Yellow);
+                ReadFileUsingBufferedStream(sourcePath, bufferSize);
+                TimeSpan timeSpan2 = stopwatch.Elapsed;
+                ColorfulMessage("\nFile Read Successfully", ConsoleColor.Green);
+                ColorfulMessage($"\nTime Taken by File Stream : {timeSpan1.TotalSeconds} seconds", ConsoleColor.Yellow);
+                ColorfulMessage($"\nTime Taken by Buffered Stream : {timeSpan2.TotalSeconds} seconds", ConsoleColor.Yellow);
+                stopwatch.Stop();
+            }
+            else if (!File.Exists(sourcePath))
```
Comment (sandeep-sol) [inline src/Assignment18/FileDataProcessor/DataProcessor.cs]: Please defer using ! in the if conditions as it will be confusing in some cases.

File: src/Assignment18/FileDataProcessor/DataProcessor.cs [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        /// Read the content from the file
+        /// </summary>
+        /// <param name="sourcePath">It is file path of the source</param>
+        /// <param name="destinationPath">It is file path of the destination</param>
+        public static void ReadFromTheFile(string sourcePath, string destinationPath)
+        {
+            if (File.Exists(sourcePath) && File.Exists(destinationPath))
+            {
+                int fileSize = (int)new FileInfo(sourcePath).Length;
+                int bufferSize = (int)Math.Min(fileSize, int.MaxValue);
```
Comment (sandeep-sol) [inline src/Assignment18/FileDataProcessor/DataProcessor.cs]: The output of Math.Min is already an integer. The typecaset here is not required.

File: src/Assignment18/FileDataProcessor/DataProcessor.cs [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        }
+
+        /// <summary>
+        /// Read the file using the FileStream
+        /// </summary>
+        /// <param name="sourcePath">It is file path of the source</param>
+        /// <param name="bufferSize">It consists of the buffer size</param>
+        public static void ReadFileUsingFileStream(string sourcePath, int bufferSize)
+        {
+            using (FileStream fileStream = new FileStream(sourcePath, FileMode.Open))
```
Comment (sandeep-sol) [inline src/Assignment18/FileDataProcessor/DataProcessor.cs:60]: You also have another way of declaring these disposable objects. What is the other way? Why have you chosen this way to use the disposable variable?

File: src/Assignment18/FileDataProcessor/DataProcessor.cs [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+        /// <summary>
+        /// Read the file using the FileStream
+        /// </summary>
+        /// <param name="sourcePath">It is file path of the source</param>
+        /// <param name="bufferSize">It is consists of buffer size</param>
+        public static void ReadFileUsingBufferedStream(string sourcePath, int bufferSize)
+        {
+            using (FileStream fileStream = new FileStream(sourcePath, FileMode.Open))
+            using (BufferedStream bufferedStream = new BufferedStream(fileStream, bufferSize))
```
Comment (sandeep-sol) [inline src/Assignment18/FileDataProcessor/DataProcessor.cs]:
```suggestion
            { using (BufferedStream bufferedStream = new BufferedStream(fileStream, bufferSize))
```

File: src/Assignment18/FileDataProcessor/DataProcessor.cs [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+            {
+                byte[] buffer = new byte[bufferSize];
+                while (fileStream.Read(buffer, 0, buffer.Length) != 0)
+                {
+                    // Read the file by the chunk
+                }
+            }
```
Comment (sandeep-sol) [inline src/Assignment18/FileDataProcessor/DataProcessor.cs:86]:
```suggestion
            } }
```

File: src/Assignment18/FileDataProcessor/DataProcessor.cs [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+        /// It will process the file and save the processed data
+        /// </summary>
+        /// <param name="sourcePath">File path of the source file</param>
+        /// <param name="destinationPath">File path of the destination file</param>
+        public static void SaveTheProcessedData(string sourcePath, string destinationPath)
+        {
+            ColorfulMessage("\nWait, To process the data in the file", ConsoleColor.Yellow);
+            if (!File.Exists(sourcePath) || !File.Exists(destinationPath))
+            {
+                ColorfulMessage("File path is doesn't exist!!!");
```
Comment (sandeep-sol) [inline src/Assignment18/FileDataProcessor/DataProcessor.cs]:
```suggestion
                ColorfulMessage("The file does not exist!!!");
```

File: src/Assignment18/FileDataProcessorAsync/FileDataProcessorAsync.cs [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        /// <param name="sourceFilePath">Path of the Source File</param>
+        /// <param name="destinationFilePath">Path of the Destination File</param>
+        /// <returns>task</returns>
+        public static async Task ProcessFileToUpperCaseAsync(string sourceFilePath, string destinationFilePath)
+        {
+            using (FileStream fileStream = new FileStream(sourceFilePath, FileMode.Open, FileAccess.Read))
+            using (BufferedStream bufferStream = new BufferedStream(fileStream))
+            using (StreamWriter streamWriter = new StreamWriter(destinationFilePath))
+            {
+                int bytesRead;
```
Comment (sandeep-sol) [inline src/Assignment18/FileDataProcessorAsync/FileDataProcessorAsync.cs]: What is the use of this variable?

File: src/Assignment18/BasicFileUsage/Program.cs [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+        /// <param name="data">Data to be added to the file</param>
+        public static void AddDataToFile(string filePath, string data)
+        {
+            byte[] buffer = Encoding.UTF8.GetBytes(data);
+            if (File.Exists(filePath))
+            {
+                // Writing to file using memoryStream
+                using (FileStream fileStream = new FileStream(filePath, FileMode.Open))
+                {
+                    using MemoryStream memoryStream = new MemoryStream();
```
Comment (sandeep-sol) [inline src/Assignment18/BasicFileUsage/Program.cs:35]: Is it necessary to write using a memory stream? Shall we directly use the File stream itself?

File: src/Assignment18/BasicFileUsage/Program.cs [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+        /// </summary>
+        /// <param name="filePath">Path at which file is present</param>
+        /// <param name="data">Data to be added to the file</param>
+        public static void AddDataToFile(string filePath, string data)
+        {
+            byte[] buffer = Encoding.UTF8.GetBytes(data);
+            if (File.Exists(filePath))
+            {
+                // Writing to file using memoryStream
+                using (FileStream fileStream = new FileStream(filePath, FileMode.Open))
```
Comment (sandeep-sol) [inline src/Assignment18/BasicFileUsage/Program.cs]: Instead of restricting the user to write the file incase of it's absence, you can use the mode `OpenOrCreate` to open the file if exists and create a file if it does not.

File: src/Assignment18/BasicFileUsage/Program.cs [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+                // Read from the file using FileStream
+                using (FileStream fileStream = new FileStream(filePath, FileMode.Open, FileAccess.Read))
+                {
+                    fileStream.Seek(0, SeekOrigin.Begin);
+                    int bytesRead;
+                    Console.WriteLine("\nAvailable file content : ");
+                    while ((bytesRead = fileStream.Read(buffer, 0, buffer.Length)) > 0)
+                    {
+                        Console.WriteLine(Encoding.ASCII.GetString(buffer, 0, bytesRead));
```
Comment (sandeep-sol) [inline src/Assignment18/BasicFileUsage/Program.cs]: Why are you encoding the data here using ASCII? Usually it will be UTF8 right?

File: tests/unit/Assignment18/BasicFileUsageTest/BasicFileUsageTestClass.cs [hunk 11]
```diff
@@ -0,0 +1,35 @@
+namespace BasicFileUsageTest
+{
+    using BasicFileUsage;
+    using System.Text;
+
+    public class BasicFileUsageTestClass
+    {
+        [Fact]
+        public void AddDataToFile_WritesAndReadsDataCorrectly()
```
Comment (sandeep-sol) [inline tests/unit/Assignment18/BasicFileUsageTest/BasicFileUsageTestClass.cs]: Update the name in the format Given_When_Then

File: tests/unit/Assignment18/BasicFileUsageTest/BasicFileUsageTestClass.cs [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+
+            // Assert
+            Assert.True(File.Exists(filePath), "File should exist after writing data.");
+
+            // Read the data from the file and check if it matches the testData
+            using (FileStream fileStream = new FileStream(filePath, FileMode.Open, FileAccess.Read))
+            {
+                byte[] buffer = new byte[1024]; // Choose an appropriate buffer size
+                int bytesRead = fileStream.Read(buffer, 0, buffer.Length);
+                string fileContent = Encoding.ASCII.GetString(buffer, 0, bytesRead);
```
Comment (sandeep-sol) [inline tests/unit/Assignment18/BasicFileUsageTest/BasicFileUsageTestClass.cs]: Why is it encoded in ASCII?

File: tests/unit/Assignment18/BasicFileUsageTest/BasicFileUsageTestClass.cs [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+            string testData = "Test data to write and read.";
+
+            // Act
+            Program.AddDataToFile(filePath, testData);
+
+            // Assert
+            Assert.True(File.Exists(filePath), "File should exist after writing data.");
+
+            // Read the data from the file and check if it matches the testData
+            using (FileStream fileStream = new FileStream(filePath, FileMode.Open, FileAccess.Read))
```
Comment (sandeep-sol) [inline tests/unit/Assignment18/BasicFileUsageTest/BasicFileUsageTestClass.cs]: Move this logic to a helper function.

File: tests/unit/Assignment18/BasicFileUsageTest/BasicFileUsageTestClass.cs
Comment (sandeep-sol) [inline tests/unit/Assignment18/BasicFileUsageTest/BasicFileUsageTestClass.cs:1]: This kind of test coverage is not enough.

You should cover all the possible test cases - positive, negative, something given wrong by the user, something wrong happens in the application, everything.

File: tests/unit/Assignment18/FileDataProcessorTest/FileDataProcessorTestClass.cs [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+        {
+            // Arrange
+            StringWriter sw = new StringWriter();
+            Console.SetOut(sw);
+            
+            // Act
+            DataProcessor.ReadFromTheFile("nonexistent.txt", TestDestinationFilePath);
+
+            // Assert
+            Assert.Contains("Source file doesn't exists", sw.ToString());
```
Comment (sandeep-sol) [inline tests/unit/Assignment18/FileDataProcessorTest/FileDataProcessorTestClass.cs]: The expected result must come from another variable.

File: tests/unit/Assignment18/FileDataProcessorTest/FileDataProcessorTestClass.cs [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+        public void SaveTheProcessedData_SourceAndDestinationFilesExist()
+        {
+            // Arrange
+            File.WriteAllText(TestSourceFilePath, "Test data to process.");
+
+            // Act
+            DataProcessor.SaveTheProcessedData(TestSourceFilePath, TestDestinationFilePath);
+
+            // Assert
+            Assert.True(File.Exists(TestSourceFilePath));
```
Comment (sandeep-sol) [inline tests/unit/Assignment18/FileDataProcessorTest/FileDataProcessorTestClass.cs:82]: For saving don't you need to compare the data written to the file? 

And also for reading, don't you need to check if it is reading the contents properly?

File: src/Assignment18/LoggingSystem/Logger.cs [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+                Console.WriteLine("Error Occurred");
+            }
+        }
+
+        /// <summary>
+        /// Read the Log File from the path
+        /// </summary>
+        public static void ReadTheLogFile()
+        {
+            if (File.Exists(_logFilePath))
```
Comment (sandeep-sol) [inline src/Assignment18/LoggingSystem/Logger.cs:62]: What if an Exception occurs while reading the file?

File: tests/unit/Assignment18/LoggingSystemTest/LoggingSystemTestClass.cs
Comment (sandeep-sol) [inline tests/unit/Assignment18/LoggingSystemTest/LoggingSystemTestClass.cs:1]: Please add test cases to check if you're handling the Exceptions properly.

File: tests/unit/Assignment18/FileDataProcessorAsyncTest/FileDataProcessorAsyncTestClass.cs
Comment (sandeep-sol) [inline tests/unit/Assignment18/FileDataProcessorAsyncTest/FileDataProcessorAsyncTestClass.cs:1]: Test cases have to be added for both the `FileProcessAsync` and the `ProcessFileToUpperCaseAsync` functions also please add test cases for negative cases too.

This test coverage is not enough.

**Review by sandeep-sol — CHANGES_REQUESTED** [review]

File: src/Assignment18/FileDataProcessor/DataProcessor.cs (hunk 3 above)
Comment (aristovinceselvanathan) [reply src/Assignment18/FileDataProcessor/DataProcessor.cs:60]: By using this method it will automatically call the dispose method of the file stream need not to be called explicitly.

File: src/Assignment18/FileDataProcessor/DataProcessor.cs (hunk 5 above)
Comment (aristovinceselvanathan) [reply src/Assignment18/FileDataProcessor/DataProcessor.cs:86]: I have updated it.

File: src/Assignment18/FileDataProcessorAsync/FileDataProcessorAsync.cs (hunk 7 above)
Comment (aristovinceselvanathan) [reply src/Assignment18/FileDataProcessorAsync/FileDataProcessorAsync.cs]: It is used to store the bytes to write int to the file by buffer using the streamwriter.

File: src/Assignment18/FileDataProcessor/DataProcessor.cs (hunk 4 above)
Comment (aristovinceselvanathan) [reply src/Assignment18/FileDataProcessor/DataProcessor.cs]: I have updated it.

File: src/Assignment18/FileDataProcessor/DataProcessor.cs (hunk 6 above)
Comment (aristovinceselvanathan) [reply src/Assignment18/FileDataProcessor/DataProcessor.cs]: I have updated it.

File: src/Assignment18/LoggingSystem/Logger.cs (hunk 16 above)
Comment (aristovinceselvanathan) [reply src/Assignment18/LoggingSystem/Logger.cs:62]: I have added the exception handling

File: src/Assignment18/FileDataProcessor/DataProcessor.cs (hunk 1 above)
Comment (aristovinceselvanathan) [reply src/Assignment18/FileDataProcessor/DataProcessor.cs]: I have updated it.

File: src/Assignment18/FileDataProcessor/DataProcessor.cs (hunk 2 above)
Comment (aristovinceselvanathan) [reply src/Assignment18/FileDataProcessor/DataProcessor.cs]: I have updated it.

File: src/Assignment18/BasicFileUsage/Program.cs (hunk 8 above)
Comment (aristovinceselvanathan) [reply src/Assignment18/BasicFileUsage/Program.cs:35]: It task used to implement all the streams to store into the file.

File: src/Assignment18/BasicFileUsage/Program.cs (hunk 9 above)
Comment (aristovinceselvanathan) [reply src/Assignment18/BasicFileUsage/Program.cs]: I have updated it.

File: tests/unit/Assignment18/BasicFileUsageTest/BasicFileUsageTestClass.cs (hunk 11 above)
Comment (aristovinceselvanathan) [reply tests/unit/Assignment18/BasicFileUsageTest/BasicFileUsageTestClass.cs]: I have updated it.

File: tests/unit/Assignment18/BasicFileUsageTest/BasicFileUsageTestClass.cs (hunk 12 above)
Comment (aristovinceselvanathan) [reply tests/unit/Assignment18/BasicFileUsageTest/BasicFileUsageTestClass.cs]: I have updated it.

File: tests/unit/Assignment18/FileDataProcessorTest/FileDataProcessorTestClass.cs (hunk 14 above)
Comment (aristovinceselvanathan) [reply tests/unit/Assignment18/FileDataProcessorTest/FileDataProcessorTestClass.cs]: I have updated it.

File: src/Assignment18/BasicFileUsage/Program.cs (hunk 10 above)
Comment (aristovinceselvanathan) [reply src/Assignment18/BasicFileUsage/Program.cs]: I have updated it.

File: tests/unit/Assignment18/BasicFileUsageTest/BasicFileUsageTestClass.cs (hunk 13 above)
Comment (aristovinceselvanathan) [reply tests/unit/Assignment18/BasicFileUsageTest/BasicFileUsageTestClass.cs]: I have moved to the helper function

File: tests/unit/Assignment18/FileDataProcessorAsyncTest/FileDataProcessorAsyncTestClass.cs
Comment (aristovinceselvanathan) [reply tests/unit/Assignment18/FileDataProcessorAsyncTest/FileDataProcessorAsyncTestClass.cs:1]: I have updated it

File: tests/unit/Assignment18/LoggingSystemTest/LoggingSystemTestClass.cs
Comment (aristovinceselvanathan) [reply tests/unit/Assignment18/LoggingSystemTest/LoggingSystemTestClass.cs:1]: I have updated it

File: tests/unit/Assignment18/FileDataProcessorTest/FileDataProcessorTestClass.cs (hunk 15 above)
Comment (aristovinceselvanathan) [reply tests/unit/Assignment18/FileDataProcessorTest/FileDataProcessorTestClass.cs:82]: I have updated it.

File: tests/unit/Assignment18/BasicFileUsageTest/BasicFileUsageTestClass.cs
Comment (aristovinceselvanathan) [reply tests/unit/Assignment18/BasicFileUsageTest/BasicFileUsageTestClass.cs:1]: I have updated it.
